python/libucx/libucx/load.py

```diff
@@ -33,19 +33,53 @@
 ]
 
 
+# Loading with RTLD_LOCAL adds the library itself to the loader's
+# loaded library cache without loading any symbols into the global
+# namespace. This allows libraries that express a dependency on
+# a library to be loaded later and successfully satisfy that dependency
+# without polluting the global symbol table with symbols from
+# that library that could conflict with symbols from other DSOs.
+PREFERRED_LOAD_FLAG = ctypes.RTLD_GLOBAL
+
+
+def _load_system_installation(soname: str):
+    """Try to dlopen() the library indicated by ``soname``
+    Raises ``OSError`` if library cannot be loaded.
+    """
+    return ctypes.CDLL(soname, PREFERRED_LOAD_FLAG)
+
+
+def _load_wheel_installation(soname: str):
+    """Try to dlopen() the library indicated by ``soname``
+    Returns ``None`` if the library cannot be loaded.
+    """
+    if os.path.isfile(lib := os.path.join(os.path.dirname(__file__), "lib", soname)):
+        return ctypes.CDLL(lib, PREFERRED_LOAD_FLAG)
+    return None
+
+
 def load_library():
-    # Dynamically load libucx.so. Prefer a system library if one is present to
-    # avoid clobbering symbols that other packages might expect, but if no
-    # other library is present use the one in the wheel.
+    """Dynamically load UCX libraries"""
+    prefer_system_installation = (
+        os.getenv("RAPIDS_LIBUCX_PREFER_SYSTEM_LIBRARY", "false").lower() != "false"
+    )
+
     libraries = []
     for lib in UCX_LIBRARIES:
-        try:
-            libucx_lib = ctypes.CDLL(lib, ctypes.RTLD_GLOBAL)
-        except OSError:
-            libucx_lib = ctypes.CDLL(
-                os.path.join(os.path.dirname(__file__), "lib", lib),
-                ctypes.RTLD_GLOBAL,
-            )
+        if prefer_system_installation:
+            # Prefer a system library if one is present to
+            # avoid clobbering symbols that other packages might expect, but if no
+            # other library is present use the one in the wheel.
+            try:
+                libucx_lib = _load_system_installation(lib)
+            except OSError:
+                libucx_lib = _load_wheel_installation(lib)
+        else:
+            # Prefer the libraries bundled in this package. If they aren't found
+            # (which might be the case in builds where the library was prebuilt
+            # before packaging the wheel), look for a system installation.
+            libucx_lib = _load_wheel_installation(lib)
```

Comment:
I don't think we need the try-catching introduced in https://github.com/rapidsai/kvikio/pull/553 and similar PRs here.

For DLFW builds, these `libucx` wheels are never even installed (they're patched out of dependency lists completely).

For RAPIDS devcontainers, there isn't something like how we build "empty" `libkvikio`, `libcudf`, etc. wheels... the `libucx` wheels are installed from PyPI (because they're in the dependency lists for e.g. `ucxx`), and but then they can effectively be ignored by setting env variable `RAPIDS_LIBUCX_PREFER_SYSTEM_LIBRARY=true` to force the use of the system UCX libraries.

Reply:
IIRC we didn't build the `libucx` wheel in DLFW because it always built and used its own ucx libraries. If that package acts as a pass-through to the system libs now, similar to the "empty" libcudf wheel, we could add it to the DLFW list and stop patching it out.

Reply:
I think you have it backwards.

The `libucx` wheels **right now** prefer the system installation of UCX and only load their own copies of those libraries as a fallback.

https://github.com/rapidsai/ucx-wheels/blob/ff1946193342bdc75c082c3d6aa153d8cdd400a9/python/libucx/libucx/load.py#L41-L48

This PR **reverses** that behavior... having them prefer their own copies of those libraries by default. It also introduces an environment variable to opt out of that and instead say "please prefer the system libraries".